### DISCUSSION TITLE
Disabling EAS

### DISCUSSION THREAD

**Discussion by diepquynh** [Joshua-Riek/ubuntu-rockchip#113] [General]
Should EAS (Energy Aware Scheduling) be disabled? EAS is a feature implemented in the core Linux kernel scheduler framework to support energy-aware task placement (e.g when to place tasks on a big or little CPU core to save energy). However, it adds a significant amount of extra latency due to expensive calculation for each task placement decision

Since this board is powered using AC power, my suggestion is to simply disable the feature by simply echoing 0 to `/proc/sys/kernel/sched_energy_aware` to switch to faster CFS (Completely Fair Scheduler) scheduling path

**Comment by Joshua-Riek** [comment]
By default, the CPU and GPU scaling governor is ondemand and causes significant performance issues. I have a systemd service that will set both to performance to counter this issue. I don't know much about the kernel, but it makes sense due to the previously mentioned scaling governor issues. 

Do you know if this has any meaningful performance impact?

**Reply by diepquynh** [reply]
Having EAS enabled does take a hit to performance. An explanation for non-nerds is that for every action you do, the task scheduler has to do an energy evaluation to decide which CPU should be chosen to complete that action, which is quite expensive as it has to measure system load of all available CPU cores with multiple load accumulation criteria and take CPU capacity into account (CPU capacity is a number defined in DTBO deciding which CPU is fast and slow)

By disabling EAS, the scheduler does not need to do such evaluation, and allows itself to evenly distribute load among CPUs, with CPU capacity taken into account. That means, the scheduler just needs to care whether a CPU core is fast enough to run a task, and if the system is under heavy load, it triggers load balance (with EAS enabled, load balance is triggered only when total system load is higher than 80%)

**Reply by diepquynh** [reply]
Explaining how EAS works in detail is actually really tough as it involves a whole bunch of stuffs related to Linux kernel scheduler, but from my experience working with smartphone kernels with EAS enabled, applying to this case is just simply disable it

**Reply by diepquynh** [reply]
Setting governor to performance does play a part, as that means the current CPU capacity of all CPU cores are capped to their maximum. But since EAS is enabled, it tries to place as many tasks to the little cluster as possible so sometimes you still see stuttering while doing simple tasks like browsing web

**Reply by diepquynh** [reply]
Looks like Rockchip have their own performance hack too
https://github.com/orangepi-xunlong/linux-orangepi/blob/orange-pi-5.10-rk3588/kernel/sched/fair.c#L6890

**Reply by Joshua-Riek** [reply]
Thank you for the explanation I think I understand. I will be happy to look into this a bit more, thanks again for the information and recommendation!

**Reply by diepquynh** [reply]
rk3588s A55 CPU capacity reference. This figure is scaled with 1024. A76 core is right below a few lines
https://github.com/orangepi-xunlong/linux-orangepi/blob/orange-pi-5.10-rk3588/arch/arm64/boot/dts/rockchip/rk3588s.dtsi#L380

**Comment by arkadijs** [comment]
Not only Rockchip kernel do these ad-hoc scheduling decisions, the kernel also overrides CPU affinity mask set from user-space via `sched_setaffinity()` / `taskset` - resetting threads bound to little cores to run on BIG. And setting `sched_energy_aware` make no difference, both via `/proc` and by kernel cmdline `kernel.sched_energy_aware=0`.

> 5.10.160-rockchip #36 SMP Wed Apr 24 03:31:38 UTC 2024 aarch64
